### PR TITLE
[#155] New table: journal_persistence_ids

### PR DESCRIPTION
**Pull request by tiagomota**
This PR introduces a new metadata table to hold `journal` specific information, as in issue #155.

Changelog:

- New table `journal_metadata` and trigger on `journal` inserts that auto-populate it.
- `journal_metadata` new table name and column names are configurable like the other tables.
- Migrations scripts that take care of creating the new table and the associated triggers/functions.
- Updates to the schemas being used on the tests, both on the `core` and `migration` artifacts.
- Use this metadata table where the following queries were being called: 
    - `JournalQueries#highestSequenceNrForPersistenceId` 
    - `ReadJounalQueries#messageQuery`
- Usage of journal metadata table is controlled through config

(These previous queries are still called in case the metadata table is empty for a specific persistence_id, acting as a fallback removing the need to actually populate the metadata beforehand in some costly migration.)

### CODE REVIEW

**Comment by tiagomota** [comment]
@lomigmegard and @mkubala 👋 

Commit 9adf419 solves the tests issues and makes the build pass. Not sure if the tests objective still remains though 👐

File: core/src/test/resources/schema/postgres/nested-partitions-schema.sql [hunk 1]
```diff
@@ -64,3 +64,62 @@ CREATE TABLE IF NOT EXISTS public.snapshot
...
 );
+
+DROP TRIGGER IF EXISTS trig_check_persistence_id_max_sequence_number ON public.journal_persistence_ids;
+DROP FUNCTION IF EXISTS public.check_persistence_id_max_sequence_number();
+DROP TRIGGER IF EXISTS trig_update_journal_persistence_ids ON public.journal;
+DROP FUNCTION IF EXISTS public.update_journal_persistence_ids();
+DROP TABLE IF EXISTS public.journal_persistence_ids;
+
+CREATE TABLE public.journal_persistence_ids(
+  id BIGSERIAL,
```
Comment (lomigmegard) [inline core/src/test/resources/schema/postgres/nested-partitions-schema.sql:75]: suggestion: use `BIGINT GENERATED ALWAYS AS IDENTITY` as this is now the recommend way I believe.

File: scripts/migration-0.6.0/6-populate-journal-persistence-ids-table.sql [hunk 2]
```diff
@@ -1 +1,31 @@
...
+BEGIN
+  j_table := schema || '.' || j_table_name;
+  jpi_table := schema || '.' || jpi_table_name;
+  sql := 'INSERT INTO ' || jpi_table ||
+         ' SELECT ' ||
+           j_persistence_id_column || ', ' ||
+           'max(' || j_sequence_number_column || '), ' ||
+           'max(' || j_ordering_column || '), ' ||
+           'min(' || j_ordering_column || ')' ||
+         ' FROM ' || j_table || ' GROUP BY ' || j_persistence_id_column;
```
[the remark this answers is not in the document]
Comment (tiagomota) [reply scripts/migration-0.6.0/6-populate-journal-persistence-ids-table.sql]: Removed in favour of queries that fallback to old behaviour if metadata table does not hold data for a given persistence_id.

File: migration/src/main/scala/akka/persistence/postgres/migration/journal/JournalSchema.scala [hunk 3]
```diff
@@ -110,12 +110,19 @@ private[journal] trait JournalSchema {
...
             BEGIN
               INSERT INTO #$fullTableName (#$persistenceId, #$maxSequenceNumber, #$maxOrdering, #$minOrdering)
-              VALUES (NEW.#$jPersistenceId, NEW.#$sequenceNumber, NEW.#$ordering, NEW.#$ordering)
+              VALUES (
+                NEW.#$jPersistenceId,
+                NEW.#$sequenceNumber,
+                NEW.#$ordering,
+                CASE
+                  WHEN NEW.#$sequenceNumber = 1 THEN NEW.#$ordering
+                  ELSE -1
```
Comment (lomigmegard) [inline migration/src/main/scala/akka/persistence/postgres/migration/journal/JournalSchema.scala:119]: @tiagomota I would use the `-1` market everywhere. It makes it explicit that it's not the actual `0` ordering. All the queries should work with -1 too right?

Comment (tiagomota) [reply migration/src/main/scala/akka/persistence/postgres/migration/journal/JournalSchema.scala:119]: Yup, I'm ok with that :)

File: core/src/test/scala/akka/persistence/postgres/journal/PostgresJournalSpec.scala [hunk 4]
```diff
@@ -54,6 +55,119 @@ abstract class PostgresJournalSpec(config: String, schemaType: SchemaType)
...
+      val perId = "perId-meta-1"
+      val sender = TestProbe()
+
+      // when
+      writeSingleMessage(1, perId, sender.ref, writerUuid)
+
+      // then
+      val metadataExists = db.run(metadataTable.filter(_.persistenceId === perId).exists.result).futureValue
+      metadataExists shouldBe true
+    }
```
Comment (lomigmegard) [inline core/src/test/scala/akka/persistence/postgres/journal/PostgresJournalSpec.scala:112]: @tiagomota Any way you could test the min ordering is correctly inserted in that case (when it's the first event of a persistence id) ?

File: core/src/test/scala/akka/persistence/postgres/journal/PostgresJournalSpec.scala [hunk 5]
```diff
@@ -54,6 +55,119 @@ abstract class PostgresJournalSpec(config: String, schemaType: SchemaType)
...
+      val newMaxSeqNr =
+        db.run(metadataTable.filter(_.persistenceId === perId).map(_.maxSequenceNumber).result.head).futureValue
+      val newMaxOrdering =
+        db.run(metadataTable.filter(_.persistenceId === perId).map(_.maxOrdering).result.head).futureValue
+      val newMinOrdering =
+        db.run(metadataTable.filter(_.persistenceId === perId).map(_.minOrdering).result.head).futureValue
+
+      newMaxSeqNr shouldBe prevMaxSeqNr + 1
+      newMaxOrdering shouldBe prevMaxOrdering + 1
+      newMinOrdering shouldBe prevMinOrdering
```
Comment (lomigmegard) [inline core/src/test/scala/akka/persistence/postgres/journal/PostgresJournalSpec.scala:139]: suggestion: Add a test that the `newMinOrdering/prevMinOrdering` is _not_ -1 or 0 (greater than 0)

File: core/src/test/scala/akka/persistence/postgres/journal/PostgresJournalSpec.scala [hunk 6]
```diff
@@ -54,6 +55,119 @@ abstract class PostgresJournalSpec(config: String, schemaType: SchemaType)
...
+      val newMaxSeqNr =
+        db.run(metadataTable.filter(_.persistenceId === perId).map(_.maxSequenceNumber).result.head).futureValue
+      val newMaxOrdering =
+        db.run(metadataTable.filter(_.persistenceId === perId).map(_.maxOrdering).result.head).futureValue
+      val newMinOrdering =
+        db.run(metadataTable.filter(_.persistenceId === perId).map(_.minOrdering).result.head).futureValue
+
+      newMaxSeqNr shouldBe prevMaxSeqNr + 1
+      newMaxOrdering shouldBe prevMaxOrdering + 1
+      newMinOrdering shouldBe -1
```
Comment (lomigmegard) [inline core/src/test/scala/akka/persistence/postgres/journal/PostgresJournalSpec.scala:168]: suggestion: use a constant somewhere to name this `-1` flag? (`UNSET_MIN_ORDERING`?)

File: core/src/test/scala/akka/persistence/postgres/journal/PostgresJournalSpec.scala (hunk 4 above)
Comment (tiagomota) [reply core/src/test/scala/akka/persistence/postgres/journal/PostgresJournalSpec.scala]: Yes. When it is the first insert for a persistence_id the `min_ordering` should be equal to the `max_ordering` (the journal `ordering` value is used in both columns), and it should be greater than 0.

I've added those checks to the test.

File: core/src/test/scala/akka/persistence/postgres/journal/PostgresJournalSpec.scala (hunk 5 above)
Comment (tiagomota) [reply core/src/test/scala/akka/persistence/postgres/journal/PostgresJournalSpec.scala:150]: Added this check 👍

File: core/src/test/scala/akka/persistence/postgres/journal/PostgresJournalSpec.scala (hunk 6 above)
Comment (tiagomota) [reply core/src/test/scala/akka/persistence/postgres/journal/PostgresJournalSpec.scala]: The value is mainly used on the migrations that will insert the trigger and the `.sql` files we have for tests setups.

I've created a local variable close to the tests to enhance readability.

File: core/src/main/scala/akka/persistence/postgres/config/AkkaPersistenceConfig.scala [hunk 7]
```diff
@@ -49,6 +50,26 @@ class JournalTableConfiguration(config: Config) {
...
+  val maxOrdering: String = cfg.as[String]("maxOrdering", "max_ordering")
+  val minOrdering: String = cfg.as[String]("minOrdering", "min_ordering")
+
+  override def toString: String =
+    s"JournalMetadataTableColumnNames($id,$persistenceId,$maxSequenceNumber,$maxOrdering,$minOrdering)"
+}
+
+class JournalMetadataTableConfiguration(config: Config) {
+  private val cfg = config.asConfig("tables.journalMetadata")
+  val tableName: String = cfg.as[String]("tableName", "journal_metadata")
```
Comment (lomigmegard) [inline core/src/main/scala/akka/persistence/postgres/config/AkkaPersistenceConfig.scala:67]: @tiagomota typo on this line? `journal_metadata` is wrong no?

File: core/src/main/scala/akka/persistence/postgres/query/dao/ReadJournalQueries.scala [hunk 8]
```diff
@@ -27,16 +25,35 @@ class ReadJournalQueries(val readJournalConfig: ReadJournalConfig) {
...
+  private def _messagesOrderingBoundedQuery(
+      persistenceId: Rep[String],
+      fromSequenceNr: Rep[Long],
+      toSequenceNr: Rep[Long],
+      max: ConstColumn[Long],
+      minOrdering: Rep[Long],
+      maxOrdering: Rep[Long]): Query[JournalTable, JournalRow, Seq] =
     baseTableQuery()
       .filter(_.persistenceId === persistenceId)
+      .filter(_.deleted === false)
```
Comment (lomigmegard) [inline core/src/main/scala/akka/persistence/postgres/query/dao/ReadJournalQueries.scala:45]: No need for this clause, it's already added in `baseTableQuery`

File: core/src/main/scala/akka/persistence/postgres/config/AkkaPersistenceConfig.scala (hunk 7 above)
Comment (tiagomota) [reply core/src/main/scala/akka/persistence/postgres/config/AkkaPersistenceConfig.scala:67]: No? Or am I missing something? 🤔 That is intended. The config will use `journal_metadata` as the default name of the metadata table, unless another is provided.

File: core/src/main/scala/akka/persistence/postgres/query/dao/ReadJournalQueries.scala (hunk 8 above)
Comment (tiagomota) [reply core/src/main/scala/akka/persistence/postgres/query/dao/ReadJournalQueries.scala:45]: 🎯

**Review by lomigmegard — APPROVED** [review]

**Pull request merged**